### PR TITLE
GMOS and the previous Setup PR

### PR DESCRIPTION
**Pull request by profxj**
A bit of a bloated PR as I tacked on initial efforts with GMOS.

Much of this is the excellent new setup built by @kbwestfall 

My contribution (aside from GMOS-specific code) are a few bug fixes.
One of these was to actually implement the usage of `min_nsig` and
`lowest_nsig` in `arclines` which has done as much harm as good.

Passing tests on Travis, but probably failing the core Dev suite.  Will 
fuss with that tomorrow.  And definitely failing the flux parts of the 
Cooked tests, but am pretty sure that is due to problems in the Cooked files
on the Team Drive.

### CODE REVIEW

File: pypeit/spectrographs/keck_deimos.py [hunk 1]
```diff
@@ -269,6 +269,7 @@ def check_headers(self, headers):
                              '6.CCDNAME': '14-4-3',
                              '7.CCDNAME': '14-4-2',
                              '8.CCDNAME': '14-5-2' }
+        headers[0]['INSTRUME'] = headers[0]['INSTRUME'][:6] # To handle both 'fresh' and KOA data
```
Comment (freddavies) [inline pypeit/spectrographs/keck_deimos.py]: DEIMOS data fresh off the telescope (as opposed to KOA data) has a different instrument name in the header, `DEIMOS: real science mosaic CCD subsystem with PowerPC in VME crate`. This seems like a relatively inoffensive way to handle it for now, but I'm sure there's a more elegant approach...

File: pypeit/wavecalib.py [hunk 2]
```diff
@@ -134,12 +130,61 @@ def _build_wv_calib(self, method, skip_QA=False):
...
+            llist = vstack([CuI, ArI, ArII])
+            self.arcparam['llist'] = llist
+
             self.wv_calib = arc.simple_calib_driver(self.msarc, self.arcparam, self.arccen, ok_mask,
                                                     nfitpix=self.par['nfitpix'],
                                                     IDpixels=self.par['IDpixels'],
                                                     IDwaves=self.par['IDwaves'])
         elif method == 'arclines':
-            self.wv_calib = arc.calib_with_arclines(self.arcparam, self.arccen, ok_mask=ok_mask)
+            #self.wv_calib = arc.calib_with_arclines(self.arcparam, self.arccen, ok_mask=ok_mask)
```
Comment (jhennawi) [inline pypeit/wavecalib.py]: I'm not sure what you are doing with the logic here, but these are clearly GMOS specific hacks that should not be checked into the master branch.

**Comment by profxj** [comment]
All tests on Cooked/ are now passing

File: pypeit/core/arc.py [hunk 3]
```diff
@@ -193,7 +199,7 @@ def get_censpec_old(lordloc, rordloc, pixlocn, frame, nonlinear_counts=None, gen
     # Extract a rough spectrum of the arc in each slit
     msgs.info("Extracting an approximate arc spectrum at the centre of each slit")
     tordcen = None
-    maskslit = np.zeros(ordcen.shape[1], dtype=np.int)
+    maskslit = np.zeros(ordcen.shape[1], dtype=bool)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/arc.py]: ignoring

File: pypeit/core/arc.py [hunk 4]
```diff
@@ -627,7 +646,7 @@ def simple_calib_driver(msarc, aparm, censpec, ok_mask, nfitpix=5, get_poly=Fals
 
 
 def simple_calib(msarc, aparm, censpec, nfitpix=5, get_poly=False,
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/arc.py]: development, development, development.

File: pypeit/core/wavecal/autoid.py [hunk 5]
```diff
@@ -313,10 +312,11 @@ class General:
     Parameters
     ----------
     spec : ndarray
-      Extracted 1D Arc Spectrum
+      2D array of arcline spectra (npix,nslit)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/wavecal/autoid.py]: changed `npix` to `nspec`

File: pypeit/core/wavecal/defs.py [hunk 6]
```diff
@@ -49,7 +50,9 @@ def lines():
     line_dict['HeI'] = 2**7
     line_dict['OH_R24000'] = 2**8
     line_dict['OH_triplespec'] = 2**9
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/wavecal/defs.py]: yes, I do.  And I can explain them to you later.

File: pypeit/data/arc_lines/NIST/ArII_vacuum.ascii [hunk 7]
```diff
@@ -0,0 +1,1550 @@
+# 03-Oct-2018 (grabbed from NIST website by JXP)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/data/arc_lines/NIST/ArII_vacuum.ascii]: k

File: pypeit/spectrographs/keck_nires.py [hunk 8]
```diff
@@ -42,19 +40,22 @@ def __init__(self):
...
                             datasec         = '[1:2048,1:1024]',
                             oscansec        = '[1:2048,980:1024]'
-                            )
-            ]
+                            )]
         # Uses default timeunit
         # Uses default primary_hdrext
         # self.sky_file = ?
 
+    @property
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/spectrographs/keck_nires.py]: done

File: pypeit/wavecalib.py (hunk 2 above)
Comment (profxj) [reply pypeit/wavecalib.py]: there are no hacks on the `general` method which what
everything uses.  And given semi-brute is broken these do nothing.

**Comment by jhennawi** [comment]
Hi Guys, 

Dev suite is looking but better, but there is still a fitstable related failures in X-shooter and NIRES. See below. Please fix this. 

Thanks,
Joe


qsolaptop:joe% run_pypeit -v2 -o vlt_xshooter_nir.pypeit
[INFO]    :: util.py 431 parse_pypeit_file() - Loading the reduction file
[INFO]    :: util.py 456 parse_pypeit_file() - Found 8 raw data frames
[INFO]    :: util.py 468 parse_pypeit_file() - Input file loaded successfully
[INFO]    :: util.py 431 parse_pypeit_file() - Loading the reduction file
[INFO]    :: util.py 456 parse_pypeit_file() - Found 8 raw data frames
[INFO]    :: util.py 468 parse_pypeit_file() - Input file loaded successfully
Traceback (most recent call last):
  File "/Users/joe/anaconda3/bin//run_pypeit", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/joe/python/PypeIt/bin/run_pypeit", line 17, in <module>
    rtval = run_pypeit.main(args)
  File "/Users/joe/python/PypeIt/pypeit/scripts/run_pypeit.py", line 98, in main
    pypeIt.init_setup(args.pypeit_file, redux_dir, calibration_check=True)
  File "/Users/joe/python/PypeIt/pypeit/pypeit.py", line 516, in init_setup
    self._setup(self.pypeit_file, calibration_check=calibration_check)
  File "/Users/joe/python/PypeIt/pypeit/pypeit.py", line 652, in _setup
    use_header_id=use_header_id, sort_dir=sort_dir)
  File "/Users/joe/python/PypeIt/pypeit/pypeitsetup.py", line 445, in run
    self.build_fitstbl(strict=not setup_only)
  File "/Users/joe/python/PypeIt/pypeit/pypeitsetup.py", line 189, in build_fitstbl
    strict=strict)
  File "/Users/joe/python/PypeIt/pypeit/metadata.py", line 86, in __init__
    else self._build(file_list, strict=strict))
  File "/Users/joe/python/PypeIt/pypeit/metadata.py", line 179, in _build
    value = self.convert_time(value, date=idate)
  File "/Users/joe/python/PypeIt/pypeit/metadata.py", line 287, in convert_time
    in_time = date+'T'+in_time
TypeError: must be str, not float

qsolaptop:joe% run_pypeit -v2 -o keck_nires_nires.pypeit
[INFO]    :: util.py 431 parse_pypeit_file() - Loading the reduction file
[INFO]    :: util.py 456 parse_pypeit_file() - Found 7 raw data frames
[INFO]    :: util.py 468 parse_pypeit_file() - Input file loaded successfully
[INFO]    :: util.py 431 parse_pypeit_file() - Loading the reduction file
[INFO]    :: util.py 456 parse_pypeit_file() - Found 7 raw data frames
[INFO]    :: util.py 468 parse_pypeit_file() - Input file loaded successfully
Traceback (most recent call last):
  File "/Users/joe/anaconda3/bin//run_pypeit", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/joe/python/PypeIt/bin/run_pypeit", line 17, in <module>
    rtval = run_pypeit.main(args)
  File "/Users/joe/python/PypeIt/pypeit/scripts/run_pypeit.py", line 98, in main
    pypeIt.init_setup(args.pypeit_file, redux_dir, calibration_check=True)
  File "/Users/joe/python/PypeIt/pypeit/pypeit.py", line 516, in init_setup
    self._setup(self.pypeit_file, calibration_check=calibration_check)
  File "/Users/joe/python/PypeIt/pypeit/pypeit.py", line 652, in _setup
    use_header_id=use_header_id, sort_dir=sort_dir)
  File "/Users/joe/python/PypeIt/pypeit/pypeitsetup.py", line 445, in run
    self.build_fitstbl(strict=not setup_only)
  File "/Users/joe/python/PypeIt/pypeit/pypeitsetup.py", line 189, in build_fitstbl
    strict=strict)
  File "/Users/joe/python/PypeIt/pypeit/metadata.py", line 86, in __init__
    else self._build(file_list, strict=strict))
  File "/Users/joe/python/PypeIt/pypeit/metadata.py", line 179, in _build
    value = self.convert_time(value, date=idate)
  File "/Users/joe/python/PypeIt/pypeit/metadata.py", line 287, in convert_time
    in_time = date+'T'+in_time
TypeError: must be str, not float

**Comment by profxj** [comment]
Passing all of the original tests.  
Time to approve @jhennawi and @kbwestfall

**Comment by jhennawi** [comment]
I will approve it as soon as NIRES runs all the way through as it did
before this PR.

**Comment by profxj** [comment]
@jhennawi + @feigewang  -- NIRES dev ran fine on exposure one and then barfed in the
`pydl` bspline in exposure 2.  Am not exploring that further.

@kbwestfall -- Note that I've added a "skip" option for `flexure method`

**Pull request merged**